pkg/apis/flowcontrol/validation/validation.go

```diff
@@ -0,0 +1,296 @@
+	}
+	return allErrs
+}
+
+// ValidateGroupSubject validates subject of "Group" kind
+func ValidateGroupSubject(subject *flowcontrol.GroupSubject, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if len(subject.Name) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("name"), ""))
+	}
+	return allErrs
+}
+
+// ValidateFlowSchemaNonResourcePolicyRule validates non-resource policy-rule in the flow-schema.
+func ValidateFlowSchemaNonResourcePolicyRule(rule *flowcontrol.NonResourcePolicyRule, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if len(rule.Verbs) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("verbs"), "verbs must contain at least one value"))
+	}
+	if len(rule.Verbs) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("nonResourceURLs"), "nonResourceURLs must contain at least one value"))
+	}
+
+	if hasWildcard(rule.NonResourceURLs) && len(rule.NonResourceURLs) > 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("nonResourceURLs"), rule.NonResourceURLs, "if '*' is present, must not specify other non-resource URLs"))
+	}
+	return allErrs
+}
+
+// ValidateFlowSchemaResourcePolicyRule validates resource policy-rule in the flow-schema.
+func ValidateFlowSchemaResourcePolicyRule(rule *flowcontrol.ResourcePolicyRule, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if len(rule.Verbs) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("verbs"), "verbs must contain at least one value"))
+	}
+	if len(rule.APIGroups) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("apiGroups"), "resource rules must supply at least one api group"))
+	}
+	if len(rule.Resources) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("resources"), "resource rules must supply at least one resource"))
+	}
+
+	if hasWildcard(rule.Verbs) && len(rule.Verbs) > 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("verbs"), rule.Verbs, "if '*' is present, must not specify other verbs"))
+	}
+	if hasWildcard(rule.APIGroups) && len(rule.APIGroups) > 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("apiGroups"), rule.APIGroups, "if '*' is present, must not specify other api groups"))
+	}
+	if hasWildcard(rule.Resources) && len(rule.Resources) > 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("resources"), rule.Resources, "if '*' is present, must not specify other resources"))
+	}
+
+	// only supported verbs are allowed
+	if !supportedVerbs.IsSuperset(sets.NewString(rule.Verbs...)) {
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("verbs"), rule.Verbs, supportedVerbs.List()))
+	}
+
+	return allErrs
+}
+
+// ValidateFlowSchemaStatus validates status for the flow-schema.
+func ValidateFlowSchemaStatus(status *flowcontrol.FlowSchemaStatus, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	// conditions will not be validated
```

Comment:
hm, is this normal? It seems like Type at least should be non-empty.

Reply:
And we should validate that a given type only appears once in the list.

Reply:
added validation&tests for the status